rleplus/rleplus.go

```diff
@@ -0,0 +1,174 @@
+package rleplus
+
+import (
+	"encoding/binary"
+	"sort"
+
+	"github.com/filecoin-project/go-filecoin/rleplus/internal"
+)
+
+// Encode returns the RLE+ representation of the provided integers.
+// Also returned is the number of bits required by this encoding,
+// which is not necessarily on a byte boundary.
+//
+// The RLE+ spec is here: https://github.com/filecoin-project/specs/blob/master/data-structures.md#rle-bitset-encoding
+// and is described by the BNF Grammar:
+//
+//    <encoding> ::= <header> <blocks>
+//    <header> ::= <version> <bit>
+//    <version> ::= "00"
+//    <blocks> ::= <block> <blocks> | ""
+//    <block> ::= <block_single> | <block_short> | <block_long>
+//    <block_single> ::= "1"
+//    <block_short> ::= "01" <bit> <bit> <bit> <bit>
+//    <block_long> ::= "00" <unsigned_varint>
+//    <bit> ::= "0" | "1"
+//
+// The encoding is returned as a []byte, each byte packed starting with the low-order bit (LSB0)
+func Encode(ints []uint64) ([]byte, int) {
+	v := bitvector.BitVector{BytePacking: bitvector.LSB0}
+	firstBit, runs := RunLengths(ints)
+
+	// Add version
+	v.Push(0)
+	v.Push(0)
+
+	v.Push(firstBit)
+
+	for _, run := range runs {
+		switch {
+		case run == 1:
+			v.Push(1)
+		case run < 16:
+			v.Push(0)
+			v.Push(1)
+			v.Extend(byte(run), 4, bitvector.LSB0)
+		case run >= 16:
+			v.Push(0)
+			v.Push(0)
+			buf := make([]byte, 10)
+			numBytes := binary.PutUvarint(buf, run)
```

Comment:
Reading into this closer I now realize this 9 bytes requirement is pedantic on the part of multiformats spec and won't impact correctness because the multiformats spec is wrong about the golang varint implementation.

Still a good idea to impose a smaller max varint for RLE+ / total number of sector ids per miner actor.  I think limiting the max varint should happen later in a separate PR.